runtime/src/accounts_db.rs

```diff
@@ -1735,6 +1965,95 @@ pub mod tests {
         assert!(check_storage(&daccounts, 2, 31));
     }
 
+    fn assert_load_account(
+        accounts: &AccountsDB,
+        slot: Slot,
+        pubkey: Pubkey,
+        expected_lamports: u64,
+    ) {
+        let ancestors = vec![(slot, 0)].into_iter().collect();
+        let owner = Account::default().owner;
+        let (account, slot) = accounts
+            .load_slow(&ancestors, &pubkey)
+            .unwrap_or((Account::new(0, 0, &owner), slot));
+        assert_eq!((account.lamports, slot), (expected_lamports, slot));
+    }
+
+    fn reconstruct_accounts_db_via_serialization(accounts: AccountsDB, slot: Slot) -> AccountsDB {
+        let mut writer = Cursor::new(vec![]);
+        serialize_into(&mut writer, &AccountsDBSerialize::new(&accounts, slot)).unwrap();
+
+        let buf = writer.into_inner();
+        let mut reader = BufReader::new(&buf[..]);
+        let daccounts = AccountsDB::new(None);
+
+        let local_paths = {
+            let paths = daccounts.paths.read().unwrap();
+            AccountsDB::format_paths(paths.to_vec())
+        };
+
+        let copied_accounts = TempDir::new().unwrap();
+        // Simulate obtaining a copy of the AppendVecs from a tarball
+        copy_append_vecs(&accounts, copied_accounts.path()).unwrap();
+        daccounts
+            .accounts_from_stream(&mut reader, local_paths, copied_accounts.path())
+            .unwrap();
+
+        daccounts
+    }
+
+    fn purge_zero_lamport_accounts(accounts: &AccountsDB, slot: Slot) {
+        let ancestors = vec![(slot as Slot, 0)].into_iter().collect();
+        accounts.purge_zero_lamport_accounts(&ancestors);
+    }
+
+    #[test]
+    fn test_accounts_db_serialize_zero_and_free() {
+        solana_logger::setup();
+
+        let some_lamport = 223;
+        let zero_lamport = 0;
+        let no_data = 0;
+        let owner = Account::default().owner;
+
+        let account = Account::new(some_lamport, no_data, &owner);
+        let pubkey = Pubkey::new_rand();
+        let pubkey2 = Pubkey::new_rand();
+        let zero_lamport_account = Account::new(zero_lamport, no_data, &owner);
+
+        let filler_account = Account::new(some_lamport, no_data, &owner);
+        let filler_account_pubkey = Pubkey::new_rand();
+
+        let accounts = AccountsDB::new_single();
+
+        let mut current_slot = 1;
+        accounts.store(current_slot, &[(&pubkey, &account)]);
+        //accounts.store(current_slot, &[(&pubkey2, &account)]); // <= Enabling this line causes even my PR fails...
+        error!("#1: {:#?}", accounts.get_storage_entries()); // #1
```

Comment:
Steps how things to evolve!

```
#1: Add a test account with some lamports at slot1: The account data is stored into the index and new storage entry is created with slot: 0 and count: 1, status: Available.

#2: Make the test account unfunded at slot2: The account data is stored into the index and another new storage entry is created with slot: 1 and count: 1, status: Available.

#3: Fill the StorageEnty created at #2 with full of outdated accont datas: To make that happen, repeatedly store another unrelated accounts, resulting the StorageEntry with slot: 1, count: 1, status: Full and new StorageEntry containing spilled ones.

#4: Assertion of existing of zero lamport account succeeds: At this moment, the index holds the account data (lamport=0) inserted at #2;

#5: Purge the zero lamport account (Simulate call of this function from periodical snapshot creation)

#6: on master: Because #5's purge decrements the #2 StorageEntry's count to 0, it's got flushed, resulting with count: 0, status: Available
    In my PR: dead_slots are purged in purge_zero_lamport_accounts, #1's StorageEntry is also purged.)

#7: on master: StorageEntry from #1 is alive.
    in my pr: all of slots containg the test account is now refreshed as blank.
```

Reply:
@sakridge I found [these costly recalculation](https://github.com/solana-labs/solana/pull/7013/files#diff-2099c5256db4eb5975c8834af38f6456R1182) of `count` doesn't really help much.. I found [this still not-passing test scenario](https://github.com/solana-labs/solana/pull/7013/files#diff-2099c5256db4eb5975c8834af38f6456R2031).

Reply:
So, yet more correct solution would be stop doing `store.remove_account()` for zero lamport accounts without any complex things.